File: src/main/services/SecurityScanner.ts
import fs from 'fs';
import path from 'path';
import type { ModManifestDto, SecurityWarning } from '../../shared/types';
import { isInside } from './PathGuard';
import { DISABLED_EXTENSION } from '../constants';
const EXECUTABLE_EXTS = new Set(['.exe', '.dll', '.so', '.dylib', '.bat', '.cmd', '.sh', '.ps1', '.jar']);
const MN = Buffer.from([0x4d, 0x5a]);
const ELF = Buffer.from([0x7f, 0x45, 0x4c, 0x46]);
function looksExecutable(filePath: string): boolean {
    const ext = path.extname(filePath).toLowerCase();
    if (EXECUTABLE_EXTS.has(ext))
        return true;
    try {
        const buf = fs.readFileSync(filePath);
        const head = buf.subarray(0, 4);
        return buf.subarray(0, 2).equals(MN) || head.equals(ELF);
    }
    catch {
        return false;
    }
}
export function scanManifest(gameRoot: string, extractRoot: string, manifest: ModManifestDto): SecurityWarning[] {
    const warnings: SecurityWarning[] = [];
    const checkAsset = (localPath: string, destination: string | undefined, index: number): void => {
        if (destination) {
            const dest = path.resolve(gameRoot, destination);
            if (!isInside(gameRoot, dest)) {
                warnings.push({ field: `assets[${index}].Destination`, reason: 'path escapes game root' });
                return;
            }
            const localAbs = path.resolve(extractRoot, localPath);
            if (fs.existsSync(localAbs) && fs.statSync(localAbs).isFile() && looksExecutable(localAbs)) {
                warnings.push({ field: `assets[${index}].LocalPath`, reason: 'asset looks like an executable' });
            }
        }
    };
    manifest.assets.forEach((a, i) => checkAsset(a.localPath, a.destination, i));
    return warnings;
}

File: RBBPMM.Wpf/RBBPMM.Core/SecurityScanner.cs
namespace RBBPMM.Core;

/// <summary>
/// Lightweight static analysis of an extracted mod manifest.
/// Ported from src/main/services/SecurityScanner.ts.
/// </summary>
public static class SecurityScanner
{
    private static readonly HashSet<string> ExecutableExts =
        [".exe", ".dll", ".so", ".dylib", ".bat", ".cmd", ".sh", ".ps1", ".jar"];

    private static bool LooksExecutable(string filePath)
    {
        var ext = Path.GetExtension(filePath).ToLowerInvariant();
        if (ExecutableExts.Contains(ext))
            return true;
        try
        {
            var head = File.ReadAllBytes(filePath);
            if (head.Length >= 2 && head[0] == 0x4D && head[1] == 0x5A)
                return true; // PE (MZ)
            if (head.Length >= 4 && head[0] == 0x7F && head[1] == 0x45 && head[2] == 0x4C && head[3] == 0x46)
                return true; // ELF
        }
        catch
        {
            // ignore
        }
        return false;
    }

    public static List<SecurityWarning> ScanManifest(string gameRoot, string extractRoot, ModManifest manifest)
    {
        var warnings = new List<SecurityWarning>();
        var i = 0;
        foreach (var asset in manifest.Assets)
        {
            if (string.IsNullOrEmpty(asset.Destination))
                continue;
            var dest = Path.GetFullPath(Path.Combine(gameRoot, asset.Destination));
            if (!PathGuard.IsInside(Path.GetFullPath(gameRoot), dest))
            {
                warnings.Add(new SecurityWarning($"assets[{i}].Destination", "path escapes game root"));
                continue;
            }
            var localAbs = Path.GetFullPath(Path.Combine(extractRoot, NormalizeRel(asset.LocalPath)));
            if (File.Exists(localAbs) && LooksExecutable(localAbs))
                warnings.Add(new SecurityWarning($"assets[{i}].LocalPath", "asset looks like an executable"));
            i++;
        }
        return warnings;
    }

    public static string NormalizeRel(string p)
    {
        var n = p.Replace('\\', '/');
        return n.StartsWith('/') ? n[1..] : n;
    }
}
